src/TodosContext.tsx

```diff
@@ -0,0 +1,273 @@
+  handleUpdateTodo: (todo: Todo) => Promise<boolean>;
+  handleDeleteTodo: (todo: Todo) => void;
+  handleClearComplete: () => void;
+  handleFilter: (value: Filter) => void;
+  handleClearError: () => void;
+}
+
+enum ErrorsMessages {
+  Clear = '',
+  Load = 'Unable to load todos',
+  Add = 'Unable to add a todo',
+  Update = 'Unable to update a todo',
+  Empty = 'Title should not be empty',
+  Delete = 'Unable to delete a todo',
+
+}
+
+type Props = {
+  children: React.ReactNode,
+};
+
+const USER_ID = 11478;
+
+export const TodosContext = createContext({} as TodosContextType);
+
+export const TodosContextProvider: React.FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filter, setFilter] = useState<Filter>(Filter.All);
+  const [tempTodos, setTempTodos] = useState<Todo[]>([] as Todo[]);
+  const [isTitleFieldFocused, setIsTitleFieldFocused] = useState(true);
+  const [error, setError] = useState('');
+
+  const filteredTodos: Todo[] = useMemo(() => todos.filter(({ completed }) => {
+    switch (filter) {
+      case Filter.Active:
+        return !completed;
+      case Filter.Completed:
+        return completed;
+      default:
+        return true;
+    }
+  }), [filter, todos, tempTodos]);
+
+  const timerIdRef = useRef<number>(0);
+
+  useEffect(() => {
+    if (timerIdRef.current) {
+      window.clearTimeout(timerIdRef.current);
+    }
+
+    timerIdRef.current = window.setTimeout(() => {
+      setError(ErrorsMessages.Clear);
+    }, 3000);
+
+    return () => clearTimeout(timerIdRef.current);
+  }, [error]);
+
+  const loadTodos = async () => {
+    try {
+      setError(ErrorsMessages.Clear);
+      const fetchedTodos = await getTodos(USER_ID);
+
+      setTodos(fetchedTodos);
+    } catch {
+      setError(ErrorsMessages.Load);
+    }
+  };
+
+  useEffect(() => {
+    loadTodos();
+  }, []);
+
+  const handlerTitleFieldFocused = (status: boolean) => {
+    setIsTitleFieldFocused(status);
+  };
+
+  const handleFilter = (value: Filter) => {
+    setFilter(value);
+  };
+
+  const handleClearError = () => {
+    setError(ErrorsMessages.Clear);
+  };
+
+  const isAllTodosCompleted = todos.every(todo => todo.completed);
+
+  const handleToggleAllTodos = async () => {
+    const updatedTodos = [...todos];
+    const newStatus = { completed: false };
+
+    if (!isAllTodosCompleted) {
+      newStatus.completed = true;
+    }
+
+    const todosToRequest = !isAllTodosCompleted
+      ? updatedTodos.filter(todo => !todo.completed)
+      : updatedTodos;
+
+    try {
+      setError(ErrorsMessages.Clear);
+      setTempTodos(todos);
+
+      await Promise.all(todosToRequest.map(async (todo) => {
+        const updatedTodo = await updateTodo(todo.id, newStatus) as Todo;
+        const indexOfUpdatedTodo = updatedTodos
+          .findIndex(({ id }) => id === updatedTodo.id);
+
+        updatedTodos[indexOfUpdatedTodo].completed = updatedTodo.completed;
+        setTodos(updatedTodos);
```

Comment:
Maybe we can move this logic to `then` after `Promise.all` and refactor it a bit